types/main.d.ts

```diff
@@ -0,0 +1,296 @@
+/**
+ * Specifies how to constrain a variable in the model.
+ */
+export interface IModelVariableConstraint {
+    /** The variable should be grater or equal to this value. */
+    min?: number;
+    /** The variable should be less or equal to this value. */
+    max?: number;
+    /** The variable should be equal to this value. */
+    equal?: number;
+}
+
+/**
+ * Specifies the options when solving the problem.
+ */
+export interface IModelOptions {
+    /**
+     * For MILP problems, specifies the relative tolerance of the objective,
+     * where `0` means 0% and `1` means 100%.
+     */
+    tolerance?: number;
+    /**
+     * How many milliseconds you want to allow for the solver to try
+     * and solve the model you're running.
+     */
+    timeout?: number;
+    /**
+     * Use MIR cuts.
+     * @deprecated NOT WORKING
+     */
+    useMIRCuts?: boolean;
+    /**
+     * Defaults to `true`.
+     */
+    exitOnCycles?: boolean;
+}
+
+/**
+ * Represents an LP/MILP problem.
+ * @typeparam TSolutionVar the decision variables that will be outputed to the `Solution` object.
+ * @typeparam TInternalVar the decision variables that will not be outputed to the `Solution` object.
+ */
+export interface IModel<TSolutionVar extends string = string, TInternalVar extends string = string> {
+    /** Name of the variable that will be the optimization objective. */
+    optimize: (TSolutionVar | TInternalVar);
+    /** To which direction to optimize the objective. */
+    opType: "max" | "min";
+    /**
+     * Optimization constraints.
+     * Specify an object with variable name as keys.
+    */
+    constraints: { [variable in (TSolutionVar | TInternalVar)]?: IModelVariableConstraint };
+    /**
+     * Variable identity relations.
+     * Specify an object with variable name as keys. These variables will be outputted into solution.
+     * The values of the object represents a linear combination of all the (rest of) variables.
+     * @example
+     *      ```
+     *      {
+     *          x: { x1: 10, x2: 5, x3: 2, x: 1 }       // x = 10 x1 + 5 x2 + 2 x3
```

Comment:
Is the `x: 1` here mandatory? What will happen if I use something like
```ts
{
    x: { x1: 10, x2: 5 },
// -- OR --
    x: { x1: 10, x2: 5, x: 10 },
}
```

Reply:
Sorry, been out to lunch on this one.

the `x: 1` isn't necessary, but I like to include it.

For me, its helpful; since it lets me easily put a constraint on the variable itself.

Reply:
Thanks. What if user input something like `x: 10` instead? Is this case invalid? If so, I may need to add more type constraint on this...